### PR TITLE
Turbine Tank Turbine Changes

### PR DESCRIPTION
**Pull request by saienduri**
This commit adds the implementation for turbine tank. This allows us to currently run llama, sd models, and 30 other models e2e and upload model artifacts to Azure. There currently is a folder uploaded into Azure tankturbine storage container using turbine tank if you are interested in structure and how we keep track of the version using the date + git_sha: [container_link](https://portal.azure.com/#view/Microsoft_Azure_Storage/ContainerMenuBlade/~/overview/storageAccountId/%2Fsubscriptions%2F8c190d1b-eb91-48d5-bec5-3e7cb7412e6c%2FresourceGroups%2Fpdue-nod-ai-rg%2Fproviders%2FMicrosoft.Storage%2FstorageAccounts%2Ftankturbine/path/tankturbine/etag/%220x8DC2CE680A9B29E%22/defaultEncryptionScope/%24account-encryption-key/denyEncryptionScopeOverride~/false/defaultId//publicAccessVal/Container). The torch IR for every model is uploaded when the user choses to upload using turbine tank. Accuracy against a torch run is also tested for every model we can run e2e. This allows us to keep track of the accuracy of our models and avoid regressions. There is also an option to download turbine tank model artifacts locally, so you don't have to run the models. It is setup so that it will not redownload and use cached artifacts if already there. It will also update your local artifacts if they are not up to date. Further detail can be found in the code comments.

To run turbine tank:
Build turbine like normal.
Run `python models/turbine_models/turbine_tank/run_tank.py`

I will add the corresponding nightly CI job once this lands.

### CODE REVIEW

**Review by dan-garvey — CHANGES_REQUESTED** [review]
I think the upload is something only a builder should do, so exposing it as a command line flag is probably not ideal. Additionally make sure the storage is publically READ only, privately writable.


Also, tests

**Comment by dan-garvey** [comment]
@monorimet please ensure the sd stuff meets the modularity requirements you have on the frontend

**Comment by saienduri** [comment]
> I think the upload is something only a builder should do, so exposing it as a command line flag is probably not ideal. Additionally make sure the storage is publically READ only, privately writable.
> 
> Also, tests

Upload is not a command line flag anymore. Storage is publicly read only. Everything is also a test now.

File: models/turbine_models/turbine_tank/tank_util.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+##################### Hugging Face Image Classification Models ###################################
+from transformers import AutoModelForImageClassification
+from transformers import AutoFeatureExtractor
+from PIL import Image
+import requests
+
+
+def preprocess_input_image(model_name):
+    # from datasets import load_dataset
```
[the remark this answers is not in the document]
Comment (saienduri) [reply models/turbine_models/turbine_tank/tank_util.py:51]: all of this code is used in test_tank.py as we use the different get methods which generate input, output, and return a HF model based on the model we are trying to run through. All the preprocessing and helper methods are necessary.

File: models/turbine_models/turbine_tank/turbine_tank.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+else:
+    WORKDIR = os.path.join(str(Path.home()), ".local/turbine_tank/")
+    print(
+        f"turbine_tank local cache is located at {WORKDIR} . You may change this by assigning the TURBINE_TANK_CACHE_DIR environment variable."
+    )
+os.makedirs(WORKDIR, exist_ok=True)
+
+storage_account_key = "XSsr+KqxBLxXzRtFv3QbbdsAxdwDGe661Q1xY4ziMRtpCazN8W6HZePi6nwud5RNLC5Y7e410abg+AStyzmX1A=="
+storage_account_name = "tankturbine"
+connection_string = "DefaultEndpointsProtocol=https;AccountName=tankturbine;AccountKey=XSsr+KqxBLxXzRtFv3QbbdsAxdwDGe661Q1xY4ziMRtpCazN8W6HZePi6nwud5RNLC5Y7e410abg+AStyzmX1A==;EndpointSuffix=core.windows.net"
```
[the remark this answers is not in the document]
Comment (saienduri) [reply models/turbine_models/turbine_tank/turbine_tank.py]: Good point, I made them environment variables that we can pass in using github secrets for our github actions workflow.

**Comment by saienduri** [comment]
By the way, the test turbine models test failing here is not due to this code. With the latest dependencies (specifically transformers), we are hitting this issue: TypeError: llama_pos_shift_attention_forward() got an unexpected keyword argument 'cache_position'. I manually ran the test models job on the main branch, and it fails as well: https://github.com/nod-ai/SHARK-Turbine/actions/runs/7996420641

**Comment by dan-garvey** [comment]
lint this when you get a chance

File: models/turbine_models/custom_models/sd_inference/vae.py [hunk 3]
```diff
@@ -113,6 +115,15 @@ def main(self, inp=AbstractTensor(*sample, dtype=torch.float32)):
 
     module_str = str(CompiledModule.get_mlir_module(inst))
     safe_name = utils.create_safe_name(hf_model_name, "-vae")
+    if upload_ir:
+        with open(f"{safe_name}.mlir", "w+") as f:
+            f.write(module_str)
+        model_name_upload = hf_model_name.replace("/", "_")
+        model_name_upload = model_name_upload + "-vae-" + variant
```
Comment (dan-garvey) [inline models/turbine_models/custom_models/sd_inference/vae.py:122]: nit: use underscores instead of dashes for consistency here and elsewhere

**Review by dan-garvey — APPROVED** [review]
small nit I probably missed on a previous pass
other than that looks good to land

File: models/turbine_models/custom_models/sd_inference/vae.py (hunk 3 above)
Comment (saienduri) [reply models/turbine_models/custom_models/sd_inference/vae.py:122]: the underscore is only used to separate the org and model name. Rest is all '-' (CompVis_stable-diffusion-v1-4-vae-decode)

**Pull request merged**